File: 07_Backend/02_Backend/04_SpringSecurity/LoginQuiz/src/main/java/org/example/loginquiz/service/advanced/GalleryService.java
package org.example.loginquiz.service.advanced;

import lombok.extern.slf4j.Slf4j;
import org.example.loginquiz.model.entity.advanced.Gallery;
import org.example.loginquiz.repository.advanced.GalleryRepository;
import org.example.loginquiz.model.entity.advanced.Gallery;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;
import org.springframework.web.servlet.support.ServletUriComponentsBuilder;

import java.util.Optional;
import java.util.UUID;


/**
 * packageName : org.example.jpaexam.service.advanced
 * fileName : GalleryService
 * author : hayj6
 * date : 2024-03-25(025)
 * description :
 * 요약 :
 * <p>
 * ===========================================================
 * DATE            AUTHOR             NOTE
 * -----------------------------------------------------------
 * 2024-03-25(025)         hayj6          최초 생성
 */
@Slf4j
@Service
public class GalleryService {
    @Autowired
    GalleryRepository galleryRepository;

    //    전체 like 검색 함수
    public Page<Gallery> findAllByGalleryTitleContaining(String galleryTitle, Pageable pageable) {
        Page<Gallery> page = galleryRepository.findAllByGalleryTitleContaining(galleryTitle, pageable);
        return page;
    }

    //    TODO : 연습 : GalleryService     : save 함수 정의
//                 GalleryControlelr
//                  저장페이지 열기      : addGallery()
//                  저장버튼 클릭시 저장 : createGallery()
//                  파일 다운로드 함수   : findDownloadGallery()
//                  jsp               : add_gallery.jsp
    public Gallery save(String uuid, String galleryTitle, MultipartFile file) {
        Gallery gallery2 = null;

        try {
            if (uuid == null) {
                String tmpUuid = UUID.randomUUID().toString().replace("-", "");

                String fileDownload = ServletUriComponentsBuilder
                        .fromCurrentContextPath()
                        .path("/advanced/gallery/")
                        .path(tmpUuid)
                        .toUriString();

                Gallery gallery = new Gallery(tmpUuid
                        , galleryTitle
                        , file.getOriginalFilename()
                        , file.getBytes()
                        , fileDownload);

                gallery2 = galleryRepository.save(gallery); // DB 저장

            } else {
                String fileDownload = ServletUriComponentsBuilder
                        .fromCurrentContextPath()
                        .path("/advanced/gallery/")
                        .path(uuid)
                        .toUriString();

                Gallery gallery = new Gallery(
                        uuid,
                        galleryTitle,
                        file.getOriginalFilename(),
                        file.getBytes(),
                        fileDownload);

                gallery2 = galleryRepository.save(gallery);
            }
        } catch (Exception e) {
            log.debug(e.getMessage());
        }
        return gallery2;
    }

    //    상세조회
    public Optional<Gallery> findById(String uuid) {
        return galleryRepository.findById(uuid);
    }

    //    TODO: 연습 : Gallery 도 FileDB 참고해서 삭제 기능을 완성하세요
    //    삭제 : 기본키(uuid)
    public boolean removeById(String uuid) {
        if (galleryRepository.existsById(uuid) == true) {
            galleryRepository.deleteById(uuid);
            return true;
        } else {
            return false;
        }
    }

//    연습 : Gallery 도 FileDB 참고해서 수정 기능을 완성하세요
}

File: 08_SI/01_SimpleDms/frontend-vue/src/services/advanced/GalleryService.js
import http from "@/utils/http-common";

class GalleryService{
    // 속성, 생성자, 함수(함수만 만들기)
    // TODO: 전체 조회(select) -> get 방식 -> @GetMapping
    getAll(galleryTitle, page, size){
        return http.get(`/advanced/gallery?galleryTitle=${galleryTitle}&page=${page}&size=${size}`);
    }     
    create(gallery, image){      // 제목, 내용, 설명이 들어간 fileDb 객체랑 이미지를 매개변수로 받는다.
        let formData = new FormData();  // form 객체
        // form 객체에 추가 : append("키", "값");       // Map 형태, Model 형태와 같은 (키, 값) 형태
        formData.append("galleryTitle", gallery.galleryTitle);
        formData.append("image", image);

        return http.post("advanced/gallery/upload"
                        , formData
                        , { headers: {
                                "Content-Type": "multipart/form-data",      // 이 형식으로 보낸다는 의미
                              },
                        });
    }
    get(uuid){
        return http.get(`/advanced/gallery/get/${uuid}`);
    }
    update(gallery, image){
        let formData = new FormData();
        formData.append("galleryTitle", gallery.galleryTitle);
        formData.append("image", image);

        return http.put(`/advanced/gallery/${gallery.uuid}`, formData, {
            headers: {
              "Content-Type": "multipart/form-data",
            },
          });
        }
        delete(uuid){
            return http.delete(`/advanced/gallery/deletion/${uuid}`);
        }
      }
export default new GalleryService();  // 객체 생성해서 export